### PR TITLE
PR vtrace return & advantage computation and unit tests

### PR DESCRIPTION
**Pull request by le-horizon**
Hey Haonan and Wei,

Everything is working fine for cart-pole, but breakout seems to be stuck in local optimum.

Some unit tests added.

Could you take a look?

Thanks,
Le

### CODE REVIEW

**Comment by hnyu** [comment]
Can you first resolve the conflicts and squash some commits?

**Comment by hnyu** [comment]
> Everything is working fine for cart-pole, but breakout seems to be stuck in local optimum.
> 

What do you mean by "local optimum"? The final performance is not as good as ac_breakout.gin?

**Comment by le-horizon** [comment]
Done squashing, all tests pass now.  Thanks!

File: alf/utils/value_ops.py [hunk 1]
```diff
@@ -271,3 +271,124 @@ def weighted_cumulative_td_fn(accumulated_td, weights_td_is_last):
...
+        fn=vs_target_minus_vs_fn,
+        elems=(weighted_discounts, tds, is_lasts),
+        initializer=tf.zeros_like(final_value),
+        reverse=True,
+        back_prop=False)
+
+    returns = (1 - is_lasts) * vs_target_minus_vs + values
+
+    next_vs_target_minus_vs = common.tensor_extend_zero(vs_target_minus_vs[1:])
+    next_vs_targets = next_vs_target_minus_vs + next_values
```
Comment (emailweixu) [inline alf/utils/value_ops.py]: this is simply returns[1:]

File: alf/examples/off_policy_ac_cart_pole_vtrace.gin [hunk 2]
```diff
@@ -0,0 +1,22 @@
+include 'off_policy_ac_cart_pole.gin'
+create_environment.num_parallel_environments=8
+
+# algorithm config
+ActorCriticLoss.use_td_lambda_return=False
+ActorCriticLoss.td_lambda=1.0
```
Comment (emailweixu) [inline alf/examples/off_policy_ac_cart_pole_vtrace.gin]: people often use a td_lambda smaller than 1.0, e.g. 0.95

File: alf/utils/value_ops.py [hunk 3]
```diff
@@ -271,3 +271,124 @@ def weighted_cumulative_td_fn(accumulated_td, weights_td_is_last):
...
+        return (1 - is_last) * (td + weighted_discount * vs_target_minus_vs)
+
+    vs_target_minus_vs = tf.scan(
+        fn=vs_target_minus_vs_fn,
+        elems=(weighted_discounts, tds, is_lasts),
+        initializer=tf.zeros_like(final_value),
+        reverse=True,
+        back_prop=False)
+
+    returns = (1 - is_lasts) * vs_target_minus_vs + values
```
Comment (emailweixu) [inline alf/utils/value_ops.py]: it seems that it's not necessary to multiply (1 - is_lasts) here, as it's already done at line 322

**Comment by emailweixu** [comment]
you can try this with on-policy training, it should give exact same result

**Comment by le-horizon** [comment]
This branch's history is not clean, causing rebasing of master to be unnecessarily tedious.  I'm opening a new branch and PR.  Will address comments there.

Thanks for taking a look guys!

File: alf/utils/value_ops.py [hunk 4]
```diff
@@ -271,3 +271,124 @@ def weighted_cumulative_td_fn(accumulated_td, weights_td_is_last):
...
+    advantages = common.tensor_extend_zero(advantages)
+
+    if not time_major:
+        returns = tf.transpose(a=returns)
+        advantages = tf.transpose(a=advantages)
+
+    return returns, advantages
+
+
+def calc_vtrace_returns_and_advantages(training_info,
```
[the remark this answers is not in the document]
Comment (le-horizon) [reply alf/utils/value_ops.py]: I thought about that, but moving it here make it slightly easier for the future if we want to apply vtrace to other losses e.g. SAC.

File: alf/utils/value_ops.py [hunk 5]
```diff
@@ -271,3 +271,124 @@ def weighted_cumulative_td_fn(accumulated_td, weights_td_is_last):
...
+        reverse=True,
+        back_prop=False)
+
+    returns = (1 - is_lasts) * vs_target_minus_vs + values
+
+    next_vs_target_minus_vs = common.tensor_extend_zero(vs_target_minus_vs[1:])
+    next_vs_targets = next_vs_target_minus_vs + next_values
+
+    # Note, advantage of last step cannot be computed, and is assumed to be 0.
+    advantages = (1 - is_lasts) * importance_ratio_clipped * (
```
[the remark this answers is not in the document]
Comment (le-horizon) [reply alf/utils/value_ops.py]: Good point.  Commented.

File: alf/utils/value_ops.py [hunk 6]
```diff
@@ -271,3 +271,124 @@ def weighted_cumulative_td_fn(accumulated_td, weights_td_is_last):
...
+        log_prob_clipping=0,
+        check_numerics=False,
+        debug_summaries=debug_summaries)
+
+    rewards = training_info.reward
+    values = value
+    discounts = training_info.discount * gamma
+    step_types = training_info.step_type
+
+    return vtrace_returns_and_advantages_impl(importance_ratio_clipped,
```
[the remark this answers is not in the document]
Comment (le-horizon) [reply alf/utils/value_ops.py]: Done

File: alf/utils/value_ops_test.py [hunk 7]
```diff
@@ -134,6 +139,231 @@ def test_generalized_advantage_estimation(self):
...
+        step_types = tf.transpose(a=step_types)
+        is_lasts = tf.cast(tf.equal(step_types, StepType.LAST), tf.float32)
+
+        # Advantage_s of vtrace + (1 - lambda) * gamma * V_s+1 == Advantage_s of GAE
+        expected_advantages -= (
+            (1 - td_lambda) * (1 - is_lasts) * common.tensor_extend_zero(
+                discounts[1:]) * common.tensor_extend_zero(values[1:]))
+        expected_advantages = tf.transpose(a=expected_advantages)
+        # self.assertAllClose(
+        #     expected_advantages, advantages, msg='advantages differ')
```
[the remark this answers is not in the document]
Comment (le-horizon) [reply alf/utils/value_ops_test.py]: These two tests when td_lambda is not 1 are failing, and I need to spend more time on them.  Let me do that now.

File: alf/utils/value_ops_test.py [hunk 8]
```diff
@@ -134,6 +139,231 @@ def test_generalized_advantage_estimation(self):
...
+            values=values,
+            step_types=step_types,
+            discounts=discounts,
+            # TODO(Le): need td_lambda version of discounted_return
+            time_major=False)
+        expected_returns = tf.transpose(a=expected_returns)
+        values = tf.transpose(a=values)
+        expected_returns = common.tensor_extend(expected_returns, values[-1])
+        expected_returns = tf.transpose(a=expected_returns)
+        # self.assertAllClose(expected_returns, returns, msg='returns differ')
```
[the remark this answers is not in the document]
Comment (le-horizon) [reply alf/utils/value_ops_test.py]: same here.

File: alf/utils/value_ops.py (hunk 1 above)
Comment (le-horizon) [reply alf/utils/value_ops.py]: Good point.  Done.

**Comment by emailweixu** [comment]
'git pull' does not work? Your code does not have any conflict with other parts.

**Comment by le-horizon** [comment]
Unfortunately there are some conflicts with TrainConfig etc..  Anyways, I'll use rebase from now on.

________________________________

**Pull request closed without merging**